### PR TITLE
API for generating burn txs in mobilecoind

### PR DESCRIPTION
**Pull request by eranrund**
### Motivation

We'd like to test `mc-mint-auditor` in CD. As part of that, we need a way to generate burn transactions.
This PR adds an API to `mobilecoind` that allows generating such transactions. A burn transaction is a standard MobileCoin transaction that uses the [recently added](https://github.com/mobilecoinfoundation/mobilecoin/pull/1862) `BurnRedemptionMemo` memo for the `TxOut` that gets sent to the burn address.

### In this PR
* The aforementioned new API and tests for it.

### Future Work
* Exercise this in CD together with `mc-mint-auditor`.

### CODE REVIEW

File: mobilecoind/src/service.rs [hunk 1]
```diff
@@ -976,6 +980,119 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
...
+
+        // Get the list of potential inputs passed to.
+        let input_list: Vec<UnspentTxOut> = request
+            .get_input_list()
+            .iter()
+            .enumerate()
+            .map(|(i, proto_utxo)| {
+                // Proto -> Rust struct conversion.
+                let utxo = UnspentTxOut::try_from(proto_utxo).map_err(|err| {
+                    rpc_internal_error("unspent_tx_out.try_from", err, &self.logger)
```
Comment (remoun) [inline mobilecoind/src/service.rs:1018]: Optional: Include the index

```suggestion
                    rpc_internal_error(format!("unspent_tx_out[{}].try_from", i), err, &self.logger)
```

File: mobilecoind/src/service.rs [hunk 2]
```diff
@@ -976,6 +980,119 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
...
+                            "mobilecoind_db.get_subaddress_id_by_utxo_id",
+                            err,
+                            &self.logger,
+                        )
+                    })?;
+
+                if subaddress_id.monitor_id != sender_monitor_id {
+                    return Err(RpcStatus::with_message(
+                        RpcStatusCode::INVALID_ARGUMENT,
+                        format!("input_list.{}", i),
```
Comment (remoun) [inline mobilecoind/src/service.rs:1044]:
```suggestion
                        format!("input_list[{}].monitor_id", i),
```

**Review by remoun — APPROVED** [review]

File: mobilecoind/src/service.rs [hunk 3]
```diff
@@ -3801,6 +3924,166 @@ mod test {
...
+        {
+            let mut request = request.clone();
+            request.set_redemption_memo(vec![5u8; BurnRedemptionMemo::MEMO_DATA_LEN + 1]);
+            assert!(client.generate_burn_redemption_tx(&request).is_err());
+        }
+
+        // Trying to burn more than we have results in an error.
+        {
+            let mut request = request.clone();
+            request.set_burn_amount(1_000_000_000_000);
```
Comment (nick-mobilecoin) [inline mobilecoind/src/service.rs:4082]: I think this is probably fine, and I'm not familiar with enough of the surrounding tests so it may already be covered.
but consider boundary conditions:
* burn_amount + fee == total  (should work)
* burn_amount + fee + 1 == total + 1 (should fail)

The invalid data length above is a good example as it just added 1 to the length, however this test case is exceeding the amount by `200000`, leaving quite a bit for off by one errors.

**Review by nick-mobilecoin — APPROVED** [review]

File: mobilecoind/src/payments.rs [hunk 4]
```diff
@@ -230,6 +232,7 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
         outlays: &[Outlay],
         opt_fee: u64,
         opt_tombstone: u64,
+        opt_memo_builder: Option<Box<dyn MemoBuilder + 'static + Send + Sync>>,
```
[the remark this answers is not in the document]
Comment (eranrund) [reply mobilecoind/src/payments.rs:235]: yup, I put it there for consistency

File: mobilecoind/src/service.rs [hunk 5]
```diff
@@ -976,6 +980,119 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
         Ok(response)
     }
 
+    fn generate_burn_redemption_tx_impl(
+        &mut self,
+        request: mc_mobilecoind_api::GenerateBurnRedemptionTxRequest,
+    ) -> Result<mc_mobilecoind_api::GenerateBurnRedemptionTxResponse, RpcStatus> {
+        // Get sender monitor id from request.
```
[the remark this answers is not in the document]
Comment (eranrund) [reply mobilecoind/src/service.rs:987]: I actually like having narration comments, it breaks up long methods into a few logical pieces (but yes, sometimes those pieces are only 1-2 lines long).

File: mobilecoind/src/payments.rs [hunk 6]
```diff
@@ -836,6 +844,7 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
         destinations: &[Outlay],
         tombstone_block: BlockIndex,
         fog_resolver_factory: &Arc<dyn Fn(&[FogUri]) -> Result<FPR, String> + Send + Sync>,
+        opt_memo_builder: Option<Box<dyn MemoBuilder + 'static + Send + Sync>>,
```
[the remark this answers is not in the document]
Comment (eranrund) [reply mobilecoind/src/payments.rs:847]: Yup, I prefixed with `opt_` since that was done in other places in this crate.

File: mobilecoind/src/service.rs (hunk 3 above)
Comment (eranrund) [reply mobilecoind/src/service.rs:4082]: No reason not to change it to an off by one test, I'll do that.

**Review by nick-mobilecoin — APPROVED** [review]
